etl/steps/data/meadow/hmd/2022-11-03/hmd_lt.py

```diff
@@ -0,0 +1,322 @@
+import os
+import re
+import tempfile
+from glob import glob
+from io import StringIO
+from typing import List
+
+import pandas as pd
+from owid import catalog
+from owid.catalog import Dataset, Table, TableMeta
+from owid.catalog.utils import underscore_table
+from owid.datautils.io import decompress_file
+from owid.walden import Catalog as WaldenCatalog
+from structlog import get_logger
+
+from etl.helpers import Names
+from etl.steps.data.converters import convert_walden_metadata
+
+log = get_logger()
+
+# naming conventions
+N = Names(__file__)
+
+
+# Files expected once Walden file is uncompressed
+FILES_EXPECTED = ["bltper_1x1", "bltper_1x10", "bltper_1x5", "bltper_5x1", "bltper_5x10", "bltper_5x5"]
+# Regular expression to extract relevant fields from file (used in make_df)
+FILE_REGEX = (
+    r"([a-zA-Z\-\s,]+), Life tables \(period {table}\), Total\tLast modified: (\d+ [a-zA-Z]{{3}} \d+);  Methods"
+    r" Protocol: v\d+ \(\d+\)\n\n((?s:.)*)"
+)
+# Dataset details from Walden
+NAMESPACE = "hmd"
+SHORT_NAME = "hmd_lt"
+VERSION_WALDEN = "2022-11-01"
+# Meadow version
+VERSION_MEADOW = "2022-11-01"
+
+
+def run(dest_dir: str) -> None:
+    """Run step."""
+    log.info("hmd_lt.start")
+
+    # Retrieve raw data from walden
+    walden_ds = WaldenCatalog().find_one(namespace=NAMESPACE, short_name=SHORT_NAME, version=VERSION_WALDEN)
+    local_file = walden_ds.ensure_downloaded()
+
+    # Create new dataset and reuse walden metadata
+    ds = init_meadow_dataset(dest_dir, walden_ds)
+
+    # Create and add tables to dataset
+    ds = create_and_add_tables_to_dataset(local_file, ds, walden_ds)
+
+    # Save the dataset
+    ds.save()
+    log.info("hmd_lt.end")
+
+
+def init_meadow_dataset(dest_dir: str, walden_ds: WaldenCatalog) -> Dataset:
+    """Initialize meadow dataset."""
+    ds = Dataset.create_empty(dest_dir)
+    ds.metadata = convert_walden_metadata(walden_ds)
+    ds.metadata.version = VERSION_MEADOW
+    return ds
+
+
+def create_and_add_tables_to_dataset(local_file: str, ds: Dataset, walden_ds: WaldenCatalog) -> Dataset:
+    """Create and add tables to dataset.
+
+    This method creates tables for all the folders found once `local_file` is uncompressed. Then,
+    it cleans and ads them to the dataset `ds`. It uses the metadata from `walden_ds` to create the tables.
+
+    Parameters
+    ----------
+    local_file : str
+        File to walden raw file.
+    ds : Dataset
+        Dataset where tables should be added.
+    walden_ds : WaldenCatalog
+        Walden dataset.
+
+    Returns
+    -------
+    Dataset
+        Dataset with tables
+    """
+    with tempfile.TemporaryDirectory() as tmp_dir:
+        # Decompress files
+        decompress_file(local_file, tmp_dir)
+
+        # Check available files
+        _sanity_check_files(tmp_dir)
+
+        # Load data
+        for i in [1, 5]:
+            for j in [1, 5, 10]:
```

Comment:
Wouldn't it be better to simply loop over all folders in the decompressed folder? I'm thinking that in the next update these names and numbers may change.

Reply:
There is a previous step `_sanity_check_files` which ensures that all files are present. I think that should suffice.